Cleaner calculation of the simulation timestep

RemiLehe opened the pull request:
The code that calculates the simulation timestep is sometimes a bit confusing.
In particular, for PSATD, the code still looks up the FDTD algorithm (even though it is not used), and calculates the CFL for this algorithm in order to determine the timestep.

This PR cleans up the code and changes the behavior for PSATD. In particular:
- There is now a dedicated file `WarpXComputeDt.cpp` that contains the calculation of the timestep.
- For PSATD, the timestep is always `min(dx, dy, dz)/c`.
- For Cartesian FDTD, the calculation of the CFL is now a static function of the respective stencil template `struct`. The aim is to avoid that we implement a new FDTD solver and then forget to implement its CFL calculation in the function `ComputeDt`.

dpgrote commented:
Thanks for the PR! I have two questions:

Why is the CKC solver removed from two of the input files?

Can the calculations of dt for the RZ solver be moved to CylindricalYeeAlgorithm::ComputeMaxDt?

MaxThevenet commented:
I think some tests should be update. @RemiLehe should I do it and push on this branch? Or do you want to have a look?

RemiLehe commented:
OK, thanks for the reminder. I just pushed to that branch.

RemiLehe commented:
@dpgrote 
Thanks for your comments/questions and sorry for the late reply!
As you suggested, I moved the calculation of max dt to the `CylindricalYeeSolver`.
Regarding the removal of the CKC solver in some input scripts: this is because these inputs are only used in PSATD tests, and we had added the line for the CKC solver solely in order to properly calculate the timestep. With this new PR, the PSATD timestep calculation does not depend on the FDTD algorithm anymore, and so these lines can be removed.

MaxThevenet commented on Source/Evolve/WarpXComputeDt.cpp:
```diff
@@ -0,0 +1,87 @@
+#   include "FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CartesianNodalAlgorithm.H"
+#endif
+
+/**
+ * Determine the timestep of the simulation. */
```
Could this be in the .H file?

RemiLehe replied:
Yes, maybe. Do you mean `WarpX.H`?

MaxThevenet replied:
Yes

dpgrote commented on Source/Evolve/WarpXComputeDt.cpp:
```diff
@@ -0,0 +1,85 @@
+
+#endif
+
+    dt.resize(0);
```
Is there a reason that dt is resized twice?

RemiLehe replied:
I think that the first resize is to make sure that there is no previously-existing value in `dt` ; otherwise the second `resize` will preserve these pre-existing values instead of overwriting them. (Note that I mindlessly copied this part from the previously-existing code that calculated `dt` before this PR.)
By agree that this looks confusing ; I'll update the PR and add a comment.

dpgrote replied:
Ok. It doesn't seem like it is needed, but it's probably safer to keep it rather than risk breaking things. An alternative is to do `dt.clear()` which will do the same thing but makes its intent clear. Maybe @WeiqunZhang knows why it is there (since he wrote it).

dpgrote commented on Regression/WarpX-GPU-tests.ini:
```diff
@@ -388,7 +388,7 @@ numthreads = 1
 doVis = 0
 compareParticles = 0
-runtime_params = psatd.fftw_plan_measure=0
+runtime_params = psatd.fftw_plan_measure=0 warpx.cfl = 0.5773502691896258
```
I'll suggest that it'll be cleaner (and maybe less confusing) to use the default cfl and remove these, and resetting the benchmark cases. Future developers and users will see this and scratch their heads trying to figure out the purpose for the special values of cfl.

MaxThevenet replied:
I would be in favor of doing it in a subsequent PR, just for this, to separate the issues and to have valid regression tests on the changes in this PR (otherwise, these changes could break the code and we'd miss it because we reset the benchmark at the same time).

RemiLehe replied:
Agreed. In fact, my preference would be to do this separate PR after #1065 is merged.

dpgrote replied:
Sounds good.

ax3l commented on Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CartesianCKCAlgorithm.H:
```diff
@@ -90,6 +92,20 @@ struct CartesianCKCAlgorithm {
+    static amrex::Real ComputeMaxDt ( amrex::Real const * const dx ) {
+#if (defined WARPX_DIM_XZ)
+            // - In Cartesian 2D geometry: determined by the minimum cell size in all direction
+            amrex::Real deltat = std::min( dx[0], dx[1] )/PhysConst::c;
+#else
+            // - In Cartesian 3D geometry: determined by the minimum cell size in all direction
+            amrex::Real deltat = std::min( dx[0], std::min( dx[1], dx[2] ) )/PhysConst::c;
+#endif
+        return deltat;
```
If I had to nit-pick, but don't worry:
```suggestion
            amrex::Real const delta_t = std::min( dx[0], dx[1] )/PhysConst::c;
#else
            // - In Cartesian 3D geometry: determined by the minimum cell size in all direction
            amrex::Real const delta_t = std::min( dx[0], std::min( dx[1], dx[2] ) ) / PhysConst::c;
#endif
        return delta_t;
```

ax3l commented on Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CartesianNodalAlgorithm.H:
```diff
@@ -37,6 +39,17 @@ struct CartesianNodalAlgorithm {
+     * Compute the maximum timestep, for which the scheme remains stable
+     * (Courant-Friedrichs-Levy limit) */
+    static amrex::Real ComputeMaxDt ( amrex::Real const * const dx ) {
+        amrex::Real deltat  = 1./( std::sqrt( AMREX_D_TERM(
+                                       1./(dx[0]*dx[0]),
+                                     + 1./(dx[1]*dx[1]),
+                                     + 1./(dx[2]*dx[2]) )) * PhysConst::c );
+        return deltat;
```
Calculating the dt in SP instead of DP with final DP cast:
```suggestion
        amrex::Real const delta_t  = 1._rt / ( std::sqrt( AMREX_D_TERM(
                                           1._rt/(dx[0]*dx[0]),
                                         + 1._rt/(dx[1]*dx[1]),
                                         + 1._rt/(dx[2]*dx[2])
                                     ) ) * PhysConst::c );
        return delta_t;
```
The macro magic with the `+` is strong in this one :)

ax3l commented on Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CartesianYeeAlgorithm.H:
```diff
@@ -36,6 +38,17 @@ struct CartesianYeeAlgorithm {
+     * Compute the maximum timestep, for which the scheme remains stable
+     * (Courant-Friedrichs-Levy limit) */
+    static amrex::Real ComputeMaxDt ( amrex::Real const * const dx ) {
+        amrex::Real deltat  = 1./( std::sqrt( AMREX_D_TERM(
+                                       1./(dx[0]*dx[0]),
+                                     + 1./(dx[1]*dx[1]),
+                                     + 1./(dx[2]*dx[2]) )) * PhysConst::c );
+        return deltat;
```
```suggestion
        amrex::Real const delta_t  = 1._rt / ( std::sqrt( AMREX_D_TERM(
                                           1._rt / (dx[0]*dx[0]),
                                         + 1._rt / (dx[1]*dx[1]),
                                         + 1._rt / (dx[2]*dx[2])
                                     ) ) * PhysConst::c );
        return delta_t;
```

ax3l commented on Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CylindricalYeeAlgorithm.H:
```diff
@@ -33,6 +35,29 @@ struct CylindricalYeeAlgorithm {
+        // In the rz case, the Courant limit has been evaluated
+        // semi-analytically by R. Lehe, and resulted in the following
+        // coefficients.
+        amrex::Real multimode_coeffs[6] = { 0.2105, 1.0, 3.5234, 8.5104, 15.5059, 24.5037 };
```
```suggestion
        amrex::Real const multimode_coeffs[6] = { 0.2105, 1.0, 3.5234, 8.5104, 15.5059, 24.5037 };
```

or modern with `#include <array>`:
```suggestion
        std::array< amrex::Real, 6 > const multimode_coeffs = {{ 0.2105, 1.0, 3.5234, 8.5104, 15.5059, 24.5037 }};
```

ax3l commented on Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CylindricalYeeAlgorithm.H:
```diff
@@ -33,6 +35,29 @@ struct CylindricalYeeAlgorithm {
+        // coefficients.
+        amrex::Real multimode_coeffs[6] = { 0.2105, 1.0, 3.5234, 8.5104, 15.5059, 24.5037 };
+        amrex::Real multimode_alpha;
+        if (n_rz_azimuthal_modes < 7) {
```
I would potentially place the `6` in a `constexpr` and use it here and in the definition of the constants.

ax3l commented on Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CylindricalYeeAlgorithm.H:
```diff
@@ -33,6 +35,29 @@ struct CylindricalYeeAlgorithm {
+          multimode_alpha = multimode_coeffs[n_rz_azimuthal_modes-1];
+        } else {
+          // Use a realistic extrapolation
+          multimode_alpha = (n_rz_azimuthal_modes - 1)*(n_rz_azimuthal_modes - 1) - 0.4;
+        }
+        amrex::Real deltat = 1./( std::sqrt(
+                                     (1.+multimode_alpha)/(dx[0]*dx[0])
+                                    + 1./(dx[1]*dx[1])) * PhysConst::c );
+        return deltat;
```
```suggestion
          multimode_alpha = (n_rz_azimuthal_modes - 1._rt)*(n_rz_azimuthal_modes - 1._rt) - 0.4_rt;
        }
        amrex::Real delta_t = 1._rt / ( std::sqrt(
                                     (1._rt + multimode_alpha) / (dx[0]*dx[0])
                                    + 1._rt / (dx[1]*dx[1])
                              ) * PhysConst::c );
        return delta_t;
```

ax3l commented on Source/WarpX.H:
```diff
@@ -279,7 +279,10 @@ public:
 
     void applyMirrors(amrex::Real time);
 
+    /**
+     * Determine the timestep of the simulation. */
```
```suggestion
    /** Determine the timestep of the simulation. */
```

ax3l commented on Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CylindricalYeeAlgorithm.H:
```diff
@@ -33,6 +35,29 @@ struct CylindricalYeeAlgorithm {
         stencil_coefs_z[0] = 1._rt/cell_size[2];  // 1./dz
     }
 
+    /**
+     * Compute the maximum timestep, for which the scheme remains stable
+     * (Courant-Friedrichs-Levy limit) */
```
```suggestion
    /** Compute the maximum, CFL-stable timestep
     *
     * Compute the maximum timestep, for which the scheme remains stable
     * under the Courant-Friedrichs-Levy limit.
     */
```

ax3l commented on Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CylindricalYeeAlgorithm.H:
```diff
@@ -42,6 +43,7 @@ struct CylindricalYeeAlgorithm {
      */
     static amrex::Real ComputeMaxDt ( amrex::Real const * const dx,
                                       int const n_rz_azimuthal_modes ) {
+        using namespace amrex::literals;
```
I think this is the only line where you need `literals` explicitly.

`amrex::literals` is an "inline" namespace of `amrex` and always pulled otherwise if you already do `using namespace amrex;`

ax3l replied:
update: and `ComputeMaxDt`

ax3l commented on Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CartesianNodalAlgorithm.H:
```diff
@@ -14,6 +14,8 @@
 
 #include <array>
```
For `std::sqrt`:
```suggestion
#include <array>
#include <cmath>
```

ax3l commented on Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CartesianYeeAlgorithm.H:
```diff
@@ -14,6 +14,8 @@
 
 #include <array>
```
For `std::sqrt`:
```suggestion
#include <array>
#include <cmath>
```

ax3l commented on Source/FieldSolver/FiniteDifferenceSolver/FiniteDifferenceAlgorithms/CylindricalYeeAlgorithm.H:
```diff
@@ -14,6 +14,8 @@
 
 #include <array>
```
```suggestion
#include <array>
#include <cmath>
```

ax3l commented:
GH action CI fails with an outdated AMReX, not changed here and fixed in #1197
